File: prep/implementation/patterns_reference_imeplementation/sliding_window/LongestSubarrayWithOnesAfterReplacement.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(m) / O(1)
public class LongestSubarrayWithOnesAfterReplacement {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        int maxOnes = 0, maxLen = 0, windowStart = 0;
        for (int windowEnd = 0; windowEnd < a.length; windowEnd++) {
          if (a[windowEnd] == 1) {
            maxOnes++;
          }

          if (windowEnd - windowStart + 1 - maxOnes > k) {
            if (a[windowStart] == 1) {
              maxOnes--;
            }
            windowStart++;
          }

          maxLen = Math.max(maxLen, windowEnd - windowStart + 1);
        }

        System.out.println(maxLen);
      }
    }
  }
}

File: prep/implementation/patterns_reference_imeplementation/sliding_window/longest_subarray_with_ones_after_replacement.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    int max_len = 0, window_start = 0;
    int max_ones_count = 0;

    for (int window_end = 0; window_end < n; window_end++) {
      if (a[window_end] == 1) {
        max_ones_count++;
      }

      if (window_end - window_start + 1 - max_ones_count > k) {
        if (a[window_start] == 1) {
          max_ones_count--;
        }
        window_start++;
      }

      max_len = max(max_len, window_end - window_start + 1);
    }

    cout << max_len << "\n";
  }

  return 0;
}